source/server/overload_manager_impl.cc

```diff
@@ -266,32 +274,54 @@ ThreadLocalOverloadState& OverloadManagerImpl::getThreadLocalOverloadState() {
-void OverloadManagerImpl::updateResourcePressure(const std::string& resource, double pressure) {
-  auto action_range = resource_to_actions_.equal_range(resource);
-  std::for_each(action_range.first, action_range.second,
-                [&](ResourceToActionMap::value_type& entry) {
-                  const std::string& action = entry.second;
-                  auto action_it = actions_.find(action);
-                  ASSERT(action_it != actions_.end());
-                  const OverloadActionState old_state = action_it->second.getState();
-                  if (action_it->second.updateResourcePressure(resource, pressure)) {
-                    const auto state = action_it->second.getState();
-
-                    if (old_state.isSaturated() != state.isSaturated()) {
-                      ENVOY_LOG(debug, "Overload action {} became {}", action,
-                                (state.isSaturated() ? "saturated" : "scaling"));
-                    }
-                    tls_->runOnAllThreads([this, action, state] {
-                      tls_->getTyped<ThreadLocalOverloadStateImpl>().setState(action, state);
-                    });
-                    auto callback_range = action_to_callbacks_.equal_range(action);
-                    std::for_each(callback_range.first, callback_range.second,
-                                  [&](ActionToCallbackMap::value_type& cb_entry) {
-                                    auto& cb = cb_entry.second;
-                                    cb.dispatcher_.post([&, state]() { cb.callback_(state); });
-                                  });
-                  }
-                });
+void OverloadManagerImpl::updateResourcePressure(const std::string& resource, double pressure,
+                                                 int flush_epoch) {
+  auto [start, end] = resource_to_actions_.equal_range(resource);
+
+  std::for_each(start, end, [&](ResourceToActionMap::value_type& entry) {
+    const std::string& action = entry.second;
+    auto action_it = actions_.find(action);
+    ASSERT(action_it != actions_.end());
+    const OverloadActionState old_state = action_it->second.getState();
+    if (action_it->second.updateResourcePressure(resource, pressure)) {
+      const auto state = action_it->second.getState();
+
+      if (old_state.isSaturated() != state.isSaturated()) {
+        ENVOY_LOG(debug, "Overload action {} became {}", action,
+                  (state.isSaturated() ? "saturated" : "scaling"));
+      }
+      state_updates_to_flush_.insert_or_assign(action, state);
+      auto [callbacks_start, callbacks_end] = action_to_callbacks_.equal_range(action);
+      std::for_each(callbacks_start, callbacks_end, [&](ActionToCallbackMap::value_type& cb_entry) {
+        callbacks_to_flush_.insert_or_assign(&cb_entry.second, state);
+      });
+    }
+  });
+
+  // Eagerly flush updates if this is the last call to updateResourcePressure expected for the
+  // current epoch.
+  --flush_awaiting_updates_;
```

Comment:
ASSERT that flush_awaiting_updates_ > 0 before this. Related, I'm a little confused about the pending flush update logic. If we don't get a flush update in an interval, and you get double updates in the next interval (not sure if this is possible) isn't this logic not right? Do we need to guard against this?

Reply:
We don't actually need the eager flush logic, I just didn't want to incur as much of a performance penalty for Envoys that aren't using high-latency updaters. It's not possible to get double updates in an interval because `Resource::update` checks `pending_update_` before calling `OverloadManagerImpl::updateResourcePressure`.

Reply:
OK can you add comments here about why this is guaranteed to be safe and the assert will always hold?